dsgrep/dsgrep.1
.\"
.\" $Id: dsgrep.1,v 1.5 2001-02-28 20:43:54 ghudson Exp $
.\"
.TH DSGREP 1 "18 Jun 1990"
.ds ]W MIT Project Athena
.SH NAME
dsgrep \- Search discuss meetings with regular expressions
.SH SYNOPSIS
dsgrep [-n n_trans] [-e search_regexp] [-f meetings_file]
[-t trans_num] [-a] [-p] [-v] [-i] [meetingname1 meetingname2 ...]
.SH DESCRIPTION
.I Dsgrep
searches discuss meetings for transactions that have titles that match the
given regular expression, or have contents that match the regular
expression.  The title of the matching transactions or their contents
can be printed.

.I Dsgrep
understands the following options:
.TP 12
.B \-n \fIn_trans\fR
specifies the number of transactions to look back from the last transaction
in each discuss meeting specified.  This defaults to 50.  This option is
ignored when the -t option is specified.
.TP
.B \-e \fIsearch_regexp\fR
This specifies the regular expression to be matched, in the style of
.IR egrep (1).
A detailed explaination of the regular expression format is
given below.  If this is omitted, it defaults to "*", matching all
transactions.
.TP
.B \-f \fImeetings_file\fR
This specifies an alternative meetings file to use as a source of meetings.
If this option is not specified, it will use the file ${HOME}/.meetings., 
.TP
.B \-t \fItrans_num\fR
Print out a specific transaction, by number.  When this option is used, the
-n option is ignored.
.TP
.B \-a
Causes
.I dsgrep
to search the text of the transactions for the regular expressions as well
as their titles.
.TP
.B \-p
Causes the text of the text of the matching transactions to be printed
instead of just the title.
.TP
.B \-v
Causes errors getting transactions to be reported.  These errors are usually
due to deleted transactions in the meetings, and are ususally not worth
seeing.
.TP
.B \-d
Causes deleted transactions to be searched as well.  Otherwise, deleted
transactions are ignored.
.TP
.B \-i
Ignores case when matching against \fIsearch_regexp\fR.
.TP
.B \fImeetings\fR
This specifies the meetings to search; if it is left off, 
.I dsgrep 
will search all the meetings.
.SH "REGULAR EXPRESSION SYNTAX"
A regular expression is zero or more \fIbranches\fR, separated by `|'.
It matches anything that matches one of the branches.
.PP
A branch is zero or more \fIpieces\fR, concatenated.
It matches a match for the first, followed by a match for the second, etc.
.PP
A piece is an \fIatom\fR possibly followed by `*', `+', or `?'.
An atom followed by `*' matches a sequence of 0 or more matches of the atom.
An atom followed by `+' matches a sequence of 1 or more matches of the atom.
An atom followed by `?' matches a match of the atom, or the null string.
.PP
An atom is a regular expression in parentheses (matching a match for the
regular expression), a \fIrange\fR (see below), `.'
(matching any single character), `^' (matching the null string at the
beginning of the input string), `$' (matching the null string at the
end of the input string), a `\e' followed by a single character (matching
that character), or a single character with no other significance
(matching that character).
.PP
A \fIrange\fR is a sequence of characters enclosed in `[]'.
It normally matches any single character from the sequence.
If the sequence begins with `^',
it matches any single character \fInot\fR from the rest of the sequence.
If two characters in the sequence are separated by `\-', this is shorthand
for the full list of ASCII characters between them
(e.g. `[0-9]' matches any decimal digit).
To include a literal `]' in the sequence, make it the first character
(following a possible `^').
To include a literal `\-', make it the first or last character.
.SH AMBIGUITY
If a regular expression could match two different parts of the input string,
it will match the one which begins earliest.
If both begin in the same place	but match different lengths, or match
the same length in different ways, life gets messier, as follows.
.PP
In general, the possibilities in a list of branches are considered in
left-to-right order, the possibilities for `*', `+', and `?' are
considered longest-first, nested constructs are considered from the
outermost in, and concatenated constructs are considered leftmost-first.
The match that will be chosen is the one that uses the earliest
possibility in the first choice that has to be made.
If there is more than one choice, the next will be made in the same manner
(earliest possibility) subject to the decision on the first choice.
And so forth.
.PP
For example, `(ab|a)b*c' could match `abc' in one of two ways.
The first choice is between `ab' and `a'; since `ab' is earlier, and does
lead to a successful overall match, it is chosen.
Since the `b' is already spoken for,
the `b*' must match its last possibility\(emthe empty string\(emsince
it must respect the earlier choice.
.PP
In the particular case where no `|'s are present and there is only one
`*', `+', or `?', the net effect is that the longest possible
match will be chosen.
So `ab*', presented with `xabbbby', will match `abbbb'.
Note that if `ab*' is tried against `xabyabbbz', it
will match `ab' just after `x', due to the begins-earliest rule.
(In effect, the decision on where to start the match is the first choice
to be made, hence subsequent choices must respect it even if this leads them
to less-preferred alternatives.)

.SH SEE ALSO
.PP
egrep(1), discuss(1)


debian/changelog
debathena-discuss (10.0.15-0debathena2) unstable; urgency=low

  * Bump debian/compat to 6
  * Package discuss server utilities (Trac: #881)
  * Cleanup man pages so they don't generate errors
  * Add missing man pages to satisfy Lintian
  * Comply with Lintian's feelings on grammar
  * Fix buffer overflow when re-adding meeting (Trac: #973)

 -- Jonathan Reed <jdreed@mit.edu>  Thu, 13 Sep 2012 17:02:56 -0400

debathena-discuss (10.0.15-0debathena1) unstable; urgency=low

  * Fix FTBFS with -Werror=format-security.

 -- Geoffrey Thomas <geofft@mit.edu>  Sat, 28 Apr 2012 20:36:38 -0700

debathena-discuss (10.0.14-0debathena1) unstable; urgency=low

  * edsc now defines and sets its own com_err hook to avoid libdsk
    confusing emacs (Trac: 603)

 -- Jonathan Reed <jdreed@mit.edu>  Sun, 25 Sep 2011 17:07:38 -0400

debathena-discuss (10.0.13-0debathena2) unstable; urgency=low

  * Flesh out mkds.8 and rmds.8 some more.

 -- Benjamin Kaduk <kaduk@mit.edu>  Sun, 03 Jul 2011 18:31:43 -0400

debathena-discuss (10.0.13-0debathena1) unstable; urgency=low

  * Update our autogoo for the 21st century.
  * Add a working distclean target.
  * Update packaging to not require daconfiscate.
  * Eliminate libds, as its only use could be replaced by libdsk.
  * Check $PAGER for a pager before falling back on a compiled-in default
    ("more" by default, but "pager" under our packaging). (Trac: #629)

 -- Evan Broder <broder@mit.edu>  Sun, 27 Jun 2010 13:02:34 -0400

debathena-discuss (10.0.12-0debathena3) unstable; urgency=low

  * Don't depend on tcsh anymore

 -- Jonathan Reed <jdreed@mit.edu>  Thu, 08 Apr 2010 22:41:16 -0400

debathena-discuss (10.0.12-0debathena2) unstable; urgency=low

  * Run dsc_setup on behalf of the user, like edsc does. (Trac #426)
  * Rewrite dsc_setup in bash, because all the cool kids use it.

 -- Jonathan Reed <jdreed@mit.edu>  Thu, 08 Apr 2010 21:03:07 -0400

debathena-discuss (10.0.12-0debathena1) unstable; urgency=low

  * Stop using the obsolete AC_TYPE_SIGNAL. (Trac: #347)
  * Add missing manpages. (Trac: #266)

 -- Evan Broder <broder@mit.edu>  Sun, 15 Nov 2009 17:35:19 -0500

debathena-discuss (10.0.11-0debathena1) unstable; urgency=low

  * Move the single-letter checks in trn_expr.y:yylex to the bottom of the
    function so that they get caught after commands that start with those
    letters. (Trac: #346)

 -- Evan Broder <broder@mit.edu>  Sat, 29 Aug 2009 17:20:43 -0400

debathena-discuss (10.0.10-0debathena2) unstable; urgency=low

  * Depend on Kerberos-enabled zephyr instead of build-depending.

 -- Evan Broder <broder@mit.edu>  Mon, 27 Jul 2009 14:38:10 -0400

debathena-discuss (10.0.10-0debathena1) unstable; urgency=low

  * Patch discuss server to work without krb4. Patch by Mitch Berger.

 -- Evan Broder <broder@mit.edu>  Thu, 23 Jul 2009 21:00:55 -0400

debathena-discuss (10.0.9-0debathena1) unstable; urgency=low

  * Fix local/dumb authentication needed for mail delivery and local
    client usage. Patch by Mitch Berger. (Trac: #274)
  * Always build without krb4.

 -- Evan Broder <broder@mit.edu>  Wed, 22 Jul 2009 22:58:40 -0400

debathena-discuss (10.0.8-0debathena1) unstable; urgency=low

  * Add one-character aliases to discuss. Patch by Ben Kaduk. (Trac: #238)
  * Squash some lintian warnings by switching from ${Source-Version} to
    ${source:Version} and ${binary:Version}
  * Use HAVE_KRB4 and HAVE_KRB5 instead of KERBEROS and KERBEROS5.
  * Assorted changes to make discuss work with any combination of krb4 and
    krb5 (including neither).
  * Always build our discuss with krb5.

 -- Evan Broder <broder@mit.edu>  Mon, 18 May 2009 23:51:52 -0400

debathena-discuss (10.0.7-0debathena3) unstable; urgency=low

  * Configure the discuss inetd service.

 -- Evan Broder <broder@mit.edu>  Thu, 05 Mar 2009 15:17:20 -0500

debathena-discuss (10.0.7-0debathena2) unstable; urgency=low

  * Change DEB_AUTO_UPDATE_AUTOCONF to 2.50, not 1.
  * Remove DEB_AUTO_UPDATE_DEBIAN_CONTROL.
  * Change "Debian-Athena Project" to "Debathena Project".
  * Remove debian/control from version control.
  * Clean up debian/copyright.

 -- Tim Abbott <tabbott@mit.edu>  Sat, 24 Jan 2009 14:31:41 -0500

debathena-discuss (10.0.7-0debathena1) unstable; urgency=low

  * Revert a small part of the OS X com_err patch from 10.0.6 in favor of
    changes to aclocal.m4
  * Remove hard-coded paths from discuss-mode, in favor of getting them
    from autoconf.
  * Remove debathena-machtype as a dependency, as it's no longer needed.

 -- Evan Broder <broder@mit.edu>  Fri, 09 Jan 2009 21:47:33 -0500

debathena-discuss (10.0.6-0debathena1) unstable; urgency=low

  [ Evan Broder ]
  * Add com_err support for delete and discuss on OS X.

  [ Anders Kaseorg ]
  * Depend emacs | emacsen instead of emacs22 | emacs21 | emacsen, so that
    the emacs metapackage is pulled in by default.

 -- Anders Kaseorg <andersk@mit.edu>  Thu, 08 Jan 2009 16:35:04 -0500

debathena-discuss (10.0.5-0debathena1) unstable; urgency=low

  * Fix some trivial typos.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 19 Aug 2008 16:35:07 -0400

debathena-discuss (10.0.4-0debathena1) unstable; urgency=low

  * Fix a crash bug in edsc on 64-bit platforms.
  * Fix date display in discuss on 64-bit platforms.

 -- Greg Hudson <ghudson@mit.edu>  Mon, 30 Jun 2008 12:50:56 -0400

debathena-discuss (10.0.3-0debathena1) unstable; urgency=low

  * Fix a crash bug in edsc (int/long mismatch in short_time).
  * Put edsc crash dumps in /tmp instead of /usr/tmp.

 -- Greg Hudson <ghudson@mit.edu>  Wed, 11 Jun 2008 19:39:38 -0400

debathena-discuss (10.0.2-0debathena1) unstable; urgency=low

  * Stop assuming that long is 32 bits.
  * Start assuming that int is 32 bits.

 -- Greg Hudson <ghudson@mit.edu>  Wed, 11 Jun 2008 16:29:35 -0400

debathena-discuss (10.0.1-0debathena1) unstable; urgency=low

  * Fix obsolete names for Kerberos library functions, which no longer
    worked on Mac OS X.  Patch from broder.


dsgrep/dsgrep.c
/*
 *
 *      Copyright (C) 1991 by the Massachusetts Institute of Technology
 *      Developed by the MIT Student Information Processing Board (SIPB).
 *      For copying information, see the file mit-copyright.h in this release.
 *
 */

#ifndef lint
#ifndef SABER
static char *RCSid = "$Id: dsgrep.c,v 1.19 2002-11-11 18:25:04 ghudson Exp $";
#endif
#endif


#define MAX_MEETINGS 128

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <sys/file.h>
#include <sys/types.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <discuss/discuss.h>
#include <regex.h>

#ifndef MAX
#define MAX(a, b)	((a) < (b) ? (b) : (a))
#endif

extern tfile unix_tfile();
extern tfile mem_tfile();

int verbose_errors;
int bsize;
char *buffer;

main(argc,argv)
     int argc;
     char *argv[];
{
  extern int optind;
  extern char *optarg;

  char *meetings_file,*homedir,*regexp_str,*var;
  name_blk *meetings,*tmp_mtg;
  mtg_info meeting_info;
  char tmp_meeting_file[MAXPATHLEN];
  int n_meetings,result,i,cur_meeting,n_to_look,j,c;
  int print_trans,flags,search_trans,search_deleted,s_trans();
  int case_insens,trans_num;
  int high,low;
  regex_t search_re;
  tfile tf;
  trn_info2 ti;
  void s_to_lower();
  int tmp_fd1,tmp_fd2;
  int using_dflt_mtgs = 0;
  struct stat statb;
  
#if defined(__APPLE__) && defined(__MACH__)
  add_error_table(&et_dsc_error_table);
#else
  initialize_dsc_error_table();
#endif

  n_to_look = 50;
  print_trans = 0;
  search_trans = 0;
  verbose_errors = 0;
  case_insens = 0;
  search_deleted = 0;
  trans_num = 0;
  meetings_file = NULL;
  regexp_str = NULL;

  while ((c = getopt(argc,argv, "n:e:f:t:apvdhi")) != EOF)
    switch(c) {
    case 'n':
      n_to_look = atoi(optarg);
      break;
    case 't':
      trans_num = atoi(optarg);
      break;
    case 'p':
      print_trans=1;
      break;
    case 'd':
      search_deleted = 1;
      break;
    case 'a':
      search_trans=1;
      bsize = 4096;
      if ((buffer = malloc((unsigned) bsize)) == NULL) {
        fprintf(stderr,"dsgrep: malloc failed\n");
	exit(1);
      }
      break;
    case 'e':
      regexp_str = optarg;
      break;
    case 'f':
      meetings_file = optarg;
      break;
    case 'v':
      verbose_errors=1;
      break;
    case 'i':
      case_insens=1;
      break;
    case '?':
    case 'h':
      fprintf(stderr,"usage: dsgrep [-n n_trans]\n");
      fprintf(stderr,"           [-e title_regexp]\n");
      fprintf(stderr,"           [-t trans_num] print out a specific trans\n");
      fprintf(stderr,"           [-a]  (search text as well as title)\n");
      fprintf(stderr,"           [-p]  (print matching transactions)\n");
      fprintf(stderr,"           [-v]  (print out verbose error messages)\n");
      fprintf(stderr,"           [-d]  (search deleted transactions as well)\n");
      fprintf(stderr,"           [-i]  (convert text to lower case before searching)\n");
      fprintf(stderr,"           [-f alt_meeting_file]\n");
      fprintf(stderr,"           [meetings]   (this must be last)\n");
      exit(1);
      break;
    }

  if (regexp_str) {
    flags = REG_NOSUB | REG_EXTENDED | (case_insens ? REG_ICASE : 0);
    if (regcomp(&search_re, regexp_str, flags) != 0) {
      fprintf(stderr,"dsgrep: Invalid regular expression %s\n",regexp_str);
      exit(1);
    }
  }

  if (meetings_file) {
    var = malloc(strlen(meetings_file) + strlen("MEETINGS=") + 1);
    if (var == NULL) {
      fprintf(stderr,"dsgrep: could not allocate memory\n");
      exit(1);
    }
    sprintf(var, "MEETINGS=%s", meetings_file);
    if (putenv(var) == -1) {
      fprintf(stderr,"dsgrep: could not add environment variable\n");
      exit(1);
    }
  }

  switch (optind - argc) {
  case 0:
    dsc_expand_mtg_set(NULL,"*",&meetings,&cur_meeting,&result);
    break;
  case 1:
    dsc_expand_mtg_set(NULL,argv[optind],&meetings,&cur_meeting,&result);
    break;
  default:
    meetings = (name_blk *)calloc((unsigned)MAX_MEETINGS,sizeof(name_blk)); 
    cur_meeting = 0;
    for(i=optind;i<argc;i++) {
      dsc_expand_mtg_set(NULL,argv[i],&tmp_mtg,&n_meetings,&result);
      if ((n_meetings == 0) && (result == 0)) {
	fprintf(stderr,"dsgrep: no such meeting %s\n",argv[i]);
	continue;
      }
      if (result != 0) {
	fprintf(stderr,"dsgrep: error expanding meeting %s: %s\n",
		argv[i],error_message(result));
	exit(1);
      }
      memcpy(&meetings[cur_meeting],tmp_mtg,sizeof(name_blk));
      cur_meeting++;
    }
  }

  if (cur_meeting == 0) {
    fprintf(stderr,"No meetings selected.\n");
    exit(1);
  }

  tf = unix_tfile(1); /*stdout */

  for(i=0;i<cur_meeting;i++) {
    dsc_get_mtg_info(&meetings[i],&meeting_info,&result);
     if (result != 0) {
       fprintf(stderr,"dsgrep: error getting meeting info for meeting '%s:%s': %s\n",
	       meetings[i].hostname, meetings[i].pathname,
	       error_message(result));
       continue;
     }
    if (trans_num != 0) {
      low = trans_num;
      high = trans_num;
    } else {
      low = meeting_info.highest-n_to_look+1; 
      low = MAX(low,1);
      high = meeting_info.highest;
    }
    for(j=low;j<=high;j++)
      {
	dsc_get_trn_info2(&meetings[i],j,&ti,&result);
	if (result != 0) {
	  if (verbose_errors)
	    fprintf(stderr,"dsgrep: error getting transaction info for %s[%d]: %s\n",
		    strrchr(meetings[i].pathname,'/')+1, j,
		    error_message(result));
	  continue;
	}
	if (!search_deleted && (ti.flags & TRN_FDELETED))
	  continue;
	if (!regexp_str || !regexec(&search_re,ti.subject,0,NULL,0) ||
	    (search_trans &&
	     s_trans(meetings[i],j,ti.num_chars,&search_re,case_insens)))  {
	  printf("%s [%d]: %s\n",
		 strrchr(meetings[i].pathname,'/')+1, j, ti.subject);
	  if (print_trans) {
	    fflush(stdout);
	    dsc_get_trn(&meetings[i],j,tf,&result);
	    if ((result != 0) && verbose_errors)
	      fprintf(stderr,"dsgrep: error getting transaction %s[%d]: %s\n",
		      strrchr(meetings[i].pathname,'/')+1, j,
		      error_message(result));
	    printf("*** End of Transaction ***\n");
	  }
	}
      }
  }
  if (using_dflt_mtgs)
    unlink(tmp_meeting_file);
  exit(0);
}

int
s_trans(nbp,trans_no,n_chars,search_re,case_insens)
     name_blk nbp;
     int trans_no,n_chars;
     regex_t *search_re;
     int case_insens;
{
  tfile tf;
  int result;
  void s_to_lower();

  if (++n_chars > bsize) {
    free(buffer);
    if ((buffer = malloc((unsigned)n_chars)) == NULL) {
      fprintf(stderr,"dsgrep: malloc failed\n");
      exit(1);
    }
    bsize = n_chars;
  }
  tf = mem_tfile(buffer,bsize);
  dsc_get_trn(&nbp,trans_no,tf,&result);
  if ((result != 0) && verbose_errors){
    fprintf(stderr,"dsgrep: error getting transation %s[%d]: %s\n",
	    strrchr(nbp.pathname,'/')+1,trans_no,
	    error_message(result));
  }
  tdestroy(tf);
  buffer[n_chars-1] = '\0';
  if (case_insens) s_to_lower(buffer);
  return(!regexec(search_re,buffer,0,NULL,0));
}

void
s_to_lower(bufp)
     char *bufp;
{
  while (*bufp != '\0') {
    if (isupper(*bufp)) *bufp = tolower(*bufp);
    bufp++;
    }
}
